File: scripts/process_ankur_students.js
#!/usr/bin/env node
/**
 * Script to process Ankur student data from Excel file and add to MJV database
 */

import ExcelJS from 'exceljs';
import fs from 'fs';
import path from 'path';
import { fileURLToPath } from 'url';

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

async function processAnkurExcel() {
  try {
    const excelFile = path.join(__dirname, '..', 'attached_assets', 'Ankur_1754753443064.xlsx');
    
    console.log(`Processing Excel file: ${excelFile}`);
    
    if (!fs.existsSync(excelFile)) {
      console.error('Excel file not found:', excelFile);
      return null;
    }

    const workbook = new ExcelJS.Workbook();
    await workbook.xlsx.readFile(excelFile);
    
    const worksheet = workbook.worksheets[0]; // First worksheet
    console.log(`Found worksheet: ${worksheet.name}`);
    console.log(`Rows: ${worksheet.rowCount}, Columns: ${worksheet.columnCount}`);
    
    // Get headers from first row
    const headers = [];
    const headerRow = worksheet.getRow(1);
    headerRow.eachCell((cell, colNumber) => {
      headers[colNumber] = cell.value?.toString().trim() || `Column${colNumber}`;
    });
    
    console.log('Headers found:', headers.filter(h => h));
    
    const students = [];
    
    // Process data rows (starting from row 2)
    for (let rowNum = 2; rowNum <= worksheet.rowCount; rowNum++) {
      const row = worksheet.getRow(rowNum);
      
      // Skip empty rows
      if (row.hasValues) {
        const rowData = {};
        row.eachCell((cell, colNumber) => {
          const header = headers[colNumber];
          if (header) {
            rowData[header] = cell.value?.toString().trim() || '';
          }
        });
        
        // Map to our student schema
        const student = {
          fullName: extractValue(rowData, ['Name', 'নাম', 'Student Name', 'Student', 'নামঃ']),
          fatherName: extractValue(rowData, ['Father Name', 'বাপেকৰ নাম', "Father's Name", 'Father']),
          motherName: extractValue(rowData, ['Mother Name', 'মাকৰ নাম', "Mother's Name", 'Mother']),
          className: 'Ankur',
          stream: '',
          rollNumber: extractValue(rowData, ['Roll No', 'Roll Number', 'ৰোল নং', 'Roll']) || `${rowNum - 1}`,
          admissionNumber: extractValue(rowData, ['Admission No', 'ভৰ্তি নং', 'Admission']) || `ANK${String(rowNum - 1).padStart(3, '0')}`,
          dateOfBirth: extractValue(rowData, ['Date of Birth', 'জন্ম তাৰিখ', 'DOB', 'Birth Date']) || '2020-01-01',
          gender: normalizeGender(extractValue(rowData, ['Gender', 'লিংগ', 'Sex'])),
          category: extractValue(rowData, ['Category', 'শ্ৰেণী', 'Caste']) || 'General',
          religion: extractValue(rowData, ['Religion', 'ধৰ্ম']) || 'Hindu',
          address: extractValue(rowData, ['Address', 'ঠিকনা', 'Village']) || 'Mahuramukh',
          phoneNumber: extractValue(rowData, ['Phone', 'ফোন', 'Contact', 'Mobile']),
          admissionDate: '2025-04-01',
          academicYear: '2025-26',
          status: 'active',
          bloodGroup: extractValue(rowData, ['Blood Group', 'তেজৰ গ্ৰুপ']),
          emergencyContact: extractValue(rowData, ['Emergency Contact', 'জৰুৰীকালীন যোগাযোগ']),
          schoolId: 3 // MJV school ID
        };
        
        // Validate required fields
        if (student.fullName && student.fullName.trim()) {
          students.push(student);
          console.log(`✓ Processed: ${student.fullName} (Roll: ${student.rollNumber})`);
        } else {
          console.log(`⚠ Skipping row ${rowNum} - missing name`);
        }
      }
    }
    
    console.log(`\nProcessed ${students.length} valid students`);
    return students;
    
  } catch (error) {
    console.error('Error processing Excel file:', error);
    return null;
  }
}

function extractValue(rowData, possibleKeys) {
  for (const key of possibleKeys) {
    const exactMatch = rowData[key];
    if (exactMatch && exactMatch.trim()) {
      return exactMatch.trim();
    }
    
    // Try partial match
    const partialMatch = Object.keys(rowData).find(k => 
      k.toLowerCase().includes(key.toLowerCase()) || 
      key.toLowerCase().includes(k.toLowerCase())
    );
    
    if (partialMatch && rowData[partialMatch] && rowData[partialMatch].trim()) {
      return rowData[partialMatch].trim();
    }
  }
  return '';
}

function normalizeGender(gender) {
  if (!gender) return 'Male';
  const g = gender.toLowerCase();
  if (g.includes('f') || g.includes('ম') || g.includes('girl')) return 'Female';
  return 'Male';
}

async function addStudentsToDatabase(students) {
  console.log(`\nAttempting to add ${students.length} students to database...`);
  
  const { default: fetch } = await import('node-fetch');
  const baseUrl = 'http://localhost:5000';
  
  let successCount = 0;
  let errorCount = 0;
  
  for (const student of students) {
    try {
      const response = await fetch(`${baseUrl}/api/students`, {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json'
        },
        body: JSON.stringify(student)
      });
      
      if (response.ok) {
        console.log(`✓ Added: ${student.fullName}`);
        successCount++;
      } else {
        const errorText = await response.text();
        console.log(`✗ Failed to add ${student.fullName}: ${errorText}`);
        errorCount++;
      }
    } catch (error) {
      console.log(`✗ Error adding ${student.fullName}: ${error.message}`);
      errorCount++;
    }
  }
  
  console.log(`\nImport Summary:`);
  console.log(`✓ Successfully added: ${successCount}`);
  console.log(`✗ Failed: ${errorCount}`);
}

async function main() {
  console.log('Ankur Student Data Processor');
  console.log('=' .repeat(40));
  
  const students = await processAnkurExcel();
  
  if (students && students.length > 0) {
    // Save to JSON for backup
    const backupFile = path.join(__dirname, 'ankur_students_backup.json');
    fs.writeFileSync(backupFile, JSON.stringify(students, null, 2), 'utf8');
    console.log(`\nBackup saved to: ${backupFile}`);
    
    // Add to database
    await addStudentsToDatabase(students);
    console.log('\nStudent import process completed!');
  } else {
    console.log('No valid student data found');
  }
}

main().catch(console.error);

File: scripts/process_ankur_students.py
#!/usr/bin/env python3
"""
Script to process Ankur student data from Excel file and add to MJV database
"""

import pandas as pd
import requests
import json
from datetime import datetime
import sys
import os

def process_ankur_excel():
    """Process the Ankur Excel file and extract student data"""
    try:
        # Read the Excel file
        excel_file = 'attached_assets/Ankur_1754753443064.xlsx'
        
        print(f"Processing Excel file: {excel_file}")
        
        # Try to read the Excel file
        try:
            df = pd.read_excel(excel_file)
        except Exception as e:
            print(f"Error reading Excel file: {e}")
            return None
            
        print(f"Found {len(df)} rows in the Excel file")
        print(f"Columns: {list(df.columns)}")
        
        # Display first few rows to understand the structure
        print("\nFirst 5 rows:")
        print(df.head())
        
        # Clean and process the data
        students = []
        
        for index, row in df.iterrows():
            try:
                # Map Excel columns to our student schema
                student_data = {
                    'fullName': str(row.get('Name', row.get('নাম', row.get('Student Name', '')))).strip(),
                    'fatherName': str(row.get('Father Name', row.get('বাপেকৰ নাম', row.get("Father's Name", '')))).strip(),
                    'motherName': str(row.get('Mother Name', row.get('মাকৰ নাম', row.get("Mother's Name", '')))).strip(),
                    'className': 'Ankur',  # All students in this file are Ankur class
                    'stream': '',  # Pre-primary doesn't have streams
                    'rollNumber': str(row.get('Roll No', row.get('Roll Number', row.get('ৰোল নং', index + 1)))),
                    'admissionNumber': str(row.get('Admission No', row.get('ভৰ্তি নং', f'ANK{index + 1:03d}'))),
                    'dateOfBirth': str(row.get('Date of Birth', row.get('জন্ম তাৰিখ', row.get('DOB', '2020-01-01')))),
                    'gender': str(row.get('Gender', row.get('লিংগ', row.get('Sex', 'Male')))).strip().title(),
                    'category': str(row.get('Category', row.get('শ্ৰেণী', 'General'))).strip(),
                    'religion': str(row.get('Religion', row.get('ধৰ্ম', 'Hindu'))).strip(),
                    'address': str(row.get('Address', row.get('ঠিকনা', 'Mahuramukh'))).strip(),
                    'phoneNumber': str(row.get('Phone', row.get('ফোন', row.get('Contact', '')))).strip(),
                    'admissionDate': '2025-04-01',  # Default admission date
                    'academicYear': '2025-26',
                    'status': 'active',
                    'bloodGroup': str(row.get('Blood Group', row.get('তেজৰ গ্ৰুপ', ''))).strip(),
                    'emergencyContact': str(row.get('Emergency Contact', row.get('জৰুৰীকালীন যোগাযোগ', ''))).strip(),
                    'schoolId': 3  # MJV school ID
                }
                
                # Clean empty strings and NaN values
                for key, value in student_data.items():
                    if pd.isna(value) or value == 'nan' or value == '':
                        if key in ['fullName', 'fatherName', 'motherName']:
                            student_data[key] = 'Unknown'
                        elif key in ['rollNumber', 'admissionNumber']:
                            student_data[key] = f'ANK{index + 1:03d}'
                        else:
                            student_data[key] = ''
                
                # Validate required fields
                if student_data['fullName'] != 'Unknown' and student_data['fullName'].strip():
                    students.append(student_data)
                    print(f"Processed student: {student_data['fullName']}")
                else:
                    print(f"Skipping row {index + 1} - missing name")
                    
            except Exception as e:
                print(f"Error processing row {index + 1}: {e}")
                continue
        
        print(f"\nProcessed {len(students)} valid students")
        return students
        
    except Exception as e:
        print(f"Error processing Excel file: {e}")
        return None

def add_students_to_database(students):
    """Add students to the database via API"""
    if not students:
        print("No students to add")
        return
    
    base_url = "http://localhost:5000"
    
    for student in students:
        try:
            response = requests.post(f"{base_url}/api/students", json=student)
            if response.status_code == 201:
                print(f"✓ Added student: {student['fullName']}")
            else:
                print(f"✗ Failed to add {student['fullName']}: {response.text}")
        except Exception as e:
            print(f"✗ Error adding {student['fullName']}: {e}")

if __name__ == "__main__":
    print("Ankur Student Data Processor")
    print("=" * 40)
    
    # Process the Excel file
    students = process_ankur_excel()
    
    if students:
        print(f"\nReady to add {len(students)} students to MJV database")
        
        # Ask for confirmation
        response = input("Do you want to proceed with adding these students? (y/n): ")
        if response.lower() == 'y':
            add_students_to_database(students)
            print("\nStudent import process completed!")
        else:
            print("Import cancelled")
            
            # Save to JSON for manual review
            with open('ankur_students_data.json', 'w', encoding='utf-8') as f:
                json.dump(students, f, indent=2, ensure_ascii=False)
            print("Student data saved to ankur_students_data.json for review")
    else:
        print("No valid student data found")